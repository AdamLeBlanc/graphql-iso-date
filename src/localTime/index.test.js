// @flow

import GraphQLLocalTime from './';
import * as Kind from 'graphql/language/kinds';

const {stringify} = require('jest-matcher-utils');

const createTime = (
  hours:number = 0,
  minutes:number = 0,
  seconds: number = 0,
  milliseconds: number = 0) => {

    const currentDate = new Date()

    return new Date(
      currentDate.getFullYear(),
      currentDate.getMonth(),
      currentDate.getDate(),
      hours,
      minutes,
      seconds,
      milliseconds
    )
  }

const invalidDates = [
  // General
  'Invalid date',
  // LocalTime with hours
  '25',
  '2Z',
  '24Z',
  // LocalTime with hours and minutes
  '24:01',
  '00:60',
  '0:60',
  '00:0',
  '00:00Z',
  '0000',
  // LocalTime with hours, minutes and seconds
  '000059',
  '00:00:60',
  '00:00:0',
  '00:00:00Z',
  // LocalTime with hours, minutes, seconds and milliseconds
  '00:00:00.1',
  '00:00:00.22',
  '00:00:00.1000',
  '00:00:00.223Z'
]

const validDates = [
  // LocalTime with hours
  [ '00', createTime() ],
  [ '13', createTime(13) ],
  [ '24', createTime(24) ],
  // LocalTime with hours and minutes
  [ '00:00', createTime() ],
  [ '23:00', createTime(23) ],
  [ '23:59', createTime(23, 59) ],
  [ '15:32', createTime(15, 32) ],
  // LocalTime with hours, minutes and seconds
  [ '00:00:00', createTime() ],
  [ '00:00:15', createTime(0, 0, 15) ],
  [ '00:00:59', createTime(0, 0, 59) ],
  // LocalTime with hours, minutes, seconds and milliseconds
  [ '00:00:00.000', createTime() ],
  [ '00:00:00.999', createTime(0, 0, 0, 999) ],
  [ '00:00:00.456', createTime(0, 0, 0, 456) ],
  [ '15:31:14.325', createTime(15, 31, 14, 325) ]
]

describe('GraphQLLocalTime', () => {

  describe('serialization', () => {
    [
      {},
      [],
      null,
      undefined,
      true,
    ].forEach(invalidInput => {
      it(`throws error when serializing ${stringify(invalidInput)}`, () => {
        expect(() =>
          GraphQLLocalTime.serialize(invalidInput)
        ).toThrowErrorMatchingSnapshot()
      })
    });

    // Serialize from Date
    [
      [ new Date(2016, 0, 1), '00:00:00.000' ],
      [ new Date(2016, 0, 1, 14, 48, 10, 3), '14:48:10.003' ],
    ].forEach(([ value, expected ]) => {
      it(`serializes javascript Date ${stringify(value)} into ${stringify(expected)}`, () => {
        expect(
          GraphQLLocalTime.serialize(value)
        ).toEqual(expected);
      })
    });

    it(`throws error when serializing invalid date`, () => {
      expect(() =>
        GraphQLLocalTime.serialize(new Date('invalid date'))
      ).toThrowErrorMatchingSnapshot();
    });

    validDates.forEach(([value]) => {
      it(`serializes date-string ${value}`, () => {
        expect(
          GraphQLLocalTime.serialize(value)
        ).toEqual(value);
      });
    });

    invalidDates.forEach(dateString => {
      it(`throws an error when serializing an invalid date-string ${stringify(dateString)}`, () => {
        expect(() =>
          GraphQLLocalTime.serialize(dateString)
        ).toThrowErrorMatchingSnapshot();
      });
    });

  });

  describe('value parsing', () => {

    validDates.forEach(([ value, expected ]) => {
      it(`parses date-string ${stringify(value)} into javascript Date ${stringify(expected)}`, () => {
        expect(
          GraphQLLocalTime.parseValue(value)
        ).toEqual(expected);
      })
    });

    [
      null,
      undefined,
      4566,
      {},
      [],
      true,
    ].forEach(invalidInput => {
      it(`throws an error when parsing ${stringify(invalidInput)}`, () => {
        expect(() =>
          GraphQLLocalTime.parseValue(invalidInput)
        ).toThrowErrorMatchingSnapshot();
      });
    });

    invalidDates.forEach(dateString => {
      it(`throws an error parsing an invalid time-string ${stringify(dateString)}`, () => {
        expect(() =>
          GraphQLLocalTime.parseValue(dateString)
        ).toThrowErrorMatchingSnapshot();
      })
    });
  });

  describe('literial parsing', () => {

    validDates.forEach(([ value, expected ]) => {
      const literal = {
        kind: Kind.STRING, value
      };

      it(`parses literal ${stringify(literal)} into javascript Date ${stringify(expected)}`, () => {
        expect(
          GraphQLLocalTime.parseLiteral(literal)
        ).toEqual(expected);
      });
    });

    invalidDates.forEach(value => {
      const invalidLiteral = {
        kind: Kind.STRING, value
      };
      it(`returns null when parsing invalid literal ${stringify(invalidLiteral)}`, () => {
        expect(
          GraphQLLocalTime.parseLiteral(invalidLiteral)
        ).toEqual(null);
      });
    });

    const invalidLiteralFloat = {
      kind: Kind.FLOAT, value: 5
    };
    it(`returns null when parsing invalid literal ${stringify(invalidLiteralFloat)}`, () => {
      expect(
        GraphQLLocalTime.parseLiteral(invalidLiteralFloat)
      ).toEqual(null);
    });
  });

});
