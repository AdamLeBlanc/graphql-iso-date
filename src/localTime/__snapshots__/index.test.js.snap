exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "0:60" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 0:60. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:0" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:0. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:0" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:0. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:00.1" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.1. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:00.22" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.22. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:00.223Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.223Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:00.1000" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.1000. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:00Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00:60" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:60. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:00Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "00:60" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:60. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "0000" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 0000. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "000059" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 000059. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "2Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 2Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "24:01" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 24:01. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "24Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 24Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "25" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 25. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws an error when serializing an invalid date-string "Invalid date" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string Invalid date. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime serialization throws error when serializing {} 1`] = `"LocalTime cannot be serialized from a non string, or non Date type {}"`;

exports[`GraphQLLocalTime serialization throws error when serializing Array [] 1`] = `"LocalTime cannot be serialized from a non string, or non Date type []"`;

exports[`GraphQLLocalTime serialization throws error when serializing invalid date 1`] = `"LocalTime cannot represent an invalid Date instance"`;

exports[`GraphQLLocalTime serialization throws error when serializing null 1`] = `"LocalTime cannot be serialized from a non string, or non Date type null"`;

exports[`GraphQLLocalTime serialization throws error when serializing true 1`] = `"LocalTime cannot be serialized from a non string, or non Date type true"`;

exports[`GraphQLLocalTime serialization throws error when serializing undefined 1`] = `"LocalTime cannot be serialized from a non string, or non Date type undefined"`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "0:60" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 0:60. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:0" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:0. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:0" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:0. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:00.1" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.1. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:00.22" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.22. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:00.223Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.223Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:00.1000" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00.1000. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:00Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:00Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00:60" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00:60. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:00Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:00Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "00:60" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 00:60. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "0000" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 0000. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "000059" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 000059. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "2Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 2Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "24:01" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 24:01. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "24Z" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 24Z. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "25" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string 25. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error parsing an invalid time-string "Invalid date" 1`] = `"LocalTime cannot represent an invalid ISO 8601 time-string Invalid date. You must provide a valid time-string in one of the following formats: HH,HH:mm,HH:mm:ss,HH:mm:ss.SSS."`;

exports[`GraphQLLocalTime value parsing throws an error when parsing {} 1`] = `"LocalTime cannot represent non string type {}"`;

exports[`GraphQLLocalTime value parsing throws an error when parsing 4566 1`] = `"LocalTime cannot represent non string type 4566"`;

exports[`GraphQLLocalTime value parsing throws an error when parsing Array [] 1`] = `"LocalTime cannot represent non string type []"`;

exports[`GraphQLLocalTime value parsing throws an error when parsing null 1`] = `"LocalTime cannot represent non string type null"`;

exports[`GraphQLLocalTime value parsing throws an error when parsing true 1`] = `"LocalTime cannot represent non string type true"`;

exports[`GraphQLLocalTime value parsing throws an error when parsing undefined 1`] = `"LocalTime cannot represent non string type undefined"`;
